[TG-1287] Parsing generic information of base classes

jeannielynnmoulton opened the pull request:
test-gen pointer update PR: https://github.com/diffblue/test-gen/pull/1401

Commit:
Defining a new type for generic bases

Commit:
Parsing generic bases' information into the class symbol

Commit:
Adding unit tests for extracting generic bases' info

thk123 commented on src/java_bytecode/java_types.h:
```diff
@@ -34,13 +34,13 @@ class java_class_typet:public class_typet
 inline const java_class_typet &to_java_class_type(const typet &type)
 {
-  assert(type.id()==ID_struct);
+  PRECONDITION(type.id()==ID_struct);
```
Nit: could this go in a separate commit

majakusber replied:
This has shown to be unnecessary (I made the change because my build was complaining, it is not anymore however). So the change was reverted.

thk123 commented on src/java_bytecode/java_types.h:
```diff
@@ -518,6 +518,82 @@ to_java_specialized_generic_class_type(typet &type)
+class java_symbol_typet : public symbol_typet
+{
+public:
+  java_symbol_typet():symbol_typet()
```
I don't think this is a useful constructor, a symbol without an identifier seems a bit meaningless, is it required?

thomasspriggs replied:
Looks like this has been removed.

majakusber replied:
See above, the class was removed.

thk123 commented on src/java_bytecode/java_types.h:
```diff
@@ -518,6 +518,82 @@ to_java_specialized_generic_class_type(typet &type)
   return static_cast<const java_specialized_generic_class_typet &>(type);
 }
 
+class java_symbol_typet : public symbol_typet
```
As it stands, this class doesn't add anything to do the code base, is there a reason to add it?

thomasspriggs replied:
Looks like this has been removed.

majakusber replied:
This class has been removed.

thk123 commented on src/java_bytecode/java_types.h:
```diff
@@ -518,6 +518,82 @@ to_java_specialized_generic_class_type(typet &type)
+    : java_symbol_typet(symbol_type)
+  {
+    set(ID_C_java_generic_symbol_type, true);
+    for(const auto &type : gen_types)
```
I think there should be a way to use `generic_types().insert` to do what you want here.

thk123 commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -145,10 +145,37 @@ void java_bytecode_convert_classt::convert(const classt &c)
+    // 'Generic<Integer>' is 'LGeneric<LInteger;>;'
+    if(
+      c.signature.has_value() &&
+      c.signature.value().substr(c.signature.value().length() - 2) == ">;")
```
When is this condition false? If never, consider making it a invariant

thk123 commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -145,10 +145,37 @@ void java_bytecode_convert_classt::convert(const classt &c)
+    // if the class extends a generic class then the class has a signature with
+    // the base class info including the generic types list at the end of it,
+    // e.g., signature for class 'A' that extends
+    // 'Generic<Integer>' is 'LGeneric<LInteger;>;'
```
So the signature for `A` contains nothing about `A`, how strange!

majakusber replied:
Yes, the signature of a class contains information about its generic parameters, superclass and interfaces but not necessarily the name of the class itself.

thk123 commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -145,10 +145,37 @@ void java_bytecode_convert_classt::convert(const classt &c)
+      c.signature.value().substr(c.signature.value().length() - 2) == ">;")
+    {
+      size_t start = 0;
+      if(c.signature.value().front() == '<')
```
So this is the generic information about `A` which we're skipping over to read `LGeneric<LInteger;>;`? Consider including this in the example in the above comment

thk123 commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -145,10 +145,37 @@ void java_bytecode_convert_classt::convert(const classt &c)
+      {
+        // the class introduces new generic parameters, the base class info
+        // starts after the closing angle bracket
+        start = find_closing_delimiter(c.signature.value(), 0, '<', '>') + 1;
```
Do we not need this information in the case? Like if I have:

```java
class B<U> {}
class A<T> extends B<T> {}
```
Do we not need to somehow annotate in the java_symbol_typet that it is an `A::T` not a `B::U`?

majakusber replied:
Yes we do, this is however done later, as part of `java_bytecode_convert_class::mark_java_implicitly_generic_class_type`

majakusber commented:
This parses the generic information of base classes. For example, for `A extends B<Integer>`, the symbol for the class `A` now shows that the base class is `B` (this functionality was there already) and it has generic type `Integer` (this is the new functionality).

thk123 commented on unit/java_bytecode/java_bytecode_parse_generics/DerivedGenerics.java:
```diff
@@ -0,0 +1,47 @@
+    // with the base class having only one type parameter.
+}
+
+class DerivedGenericUninst<T> extends Generic<T>
```
Consider here using a different letter to the letter that the class for `Generic` uses to make errors more obvious

majakusber replied:
The name of the parameter does not matter here as we are not looking at the symbol for `Generic` at all when parsing the class `DerivedGenericUninst`.

majakusber replied:
We are however testing that the parameter of the base is `java::DerivedGenericUninst::T` rather than `java::Generic::T`.

thk123 commented on unit/java_bytecode/java_bytecode_parse_generics/DerivedGenerics.java:
```diff
@@ -0,0 +1,47 @@
+    // with the base class having two type parameters, where one is specialised
+    // and the other is not.
+}
+
```
Some extra tests that might be useful:

- Classes that extend inner classes (maybe not - I guess the parsing for this is already tested)
- Inner classes that extend classes 
(As in both cases we had issues with parsing the signature)

- Classes that are specialised with a generic type (e.g. `extends Gen<ArrayList<Integer>>` (again maybe not so important as the whats after extends should already have this covered by existing tests)

- Probably something with a three hierarchy
- Something that implements an interface
- Something that implements a generic interface (both specialised and specialised)
- Something that implements a couple of interfaces
- Assuming Java allows this (hopefully not) something that just `extends Generirc` 

Sorry :blush:

majakusber commented:
After an offline discussion with @thk123, this will be redesigned - the new type extends a symbol type and symbol types are meant to refer to a particular symbol in the symbol table with the matching identifier. However, if we have something like `A<T> extends B<T,Integer>` there is no symbol for `B<T,Integer>` and there never will be (symbols are only created for `B<T,U>` and any full specialisation of the class like `B<Integer,Integer>`).

thk123 reviewed (commented):
Looks good - suggested a few more tests and had an offline discussion about some potential traps when using symbols. Sorry if some of the comments aren't relevant when considering the replacement for `generic_symbol_typet`

romainbrenguier commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -145,10 +145,37 @@ void java_bytecode_convert_classt::convert(const classt &c)
+        // the class introduces new generic parameters, the base class info
+        // starts after the closing angle bracket
+        start = find_closing_delimiter(c.signature.value(), 0, '<', '>') + 1;
+      }
```
I'd suggest using a ternary if operator and marking start const

romainbrenguier commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -145,10 +145,37 @@ void java_bytecode_convert_classt::convert(const classt &c)
+        start = find_closing_delimiter(c.signature.value(), 0, '<', '>') + 1;
+      }
+      const typet &base_type = java_type_from_string(
+        c.signature.value().substr(start, std::string::npos),
```
the second argument of `substr` is optional

majakusber replied:
This is not used anymore.

romainbrenguier commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -145,10 +145,37 @@ void java_bytecode_convert_classt::convert(const classt &c)
+        c.signature.value().substr(start, std::string::npos),
+        qualified_classname);
+      PRECONDITION(is_java_generic_type(base_type));
+      java_generic_typet gen_base_type = to_java_generic_type(base_type);
```
I would suggest defining a `can_cast_type` function for `java_generic_typet`  (should be the same as `is_java_generic_type`) and using `type_checked_cast`

majakusber replied:
We have followed the design used with other types in our code base - having a `is_.._type` and `to_.._type` functions. Would you prefer to have it changed here?

romainbrenguier commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -145,10 +145,37 @@ void java_bytecode_convert_classt::convert(const classt &c)
+      PRECONDITION(is_java_generic_type(base_type));
+      java_generic_typet gen_base_type = to_java_generic_type(base_type);
+      java_generic_symbol_typet gen_base(
+        base, gen_base_type.generic_type_arguments());
```
`gen_base` and `gen_base_type` are const

romainbrenguier commented on src/java_bytecode/java_types.h:
```diff
@@ -518,6 +518,82 @@ to_java_specialized_generic_class_type(typet &type)
+    const generic_typest &gen_types)
+    : java_symbol_typet(symbol_type)
+  {
+    set(ID_C_java_generic_symbol_type, true);
```
Why is there a `C` in this identifier? It seems java specific.

majakusber replied:
An explanation from @thk123: 
It is meant to mean comment (that is it goes into the comments section of the irep as opposed to the named subtree section). In practise the important part is the string starts with a `#`.

romainbrenguier commented on src/java_bytecode/java_types.h:
```diff
@@ -518,6 +518,82 @@ to_java_specialized_generic_class_type(typet &type)
+
+  const generic_typest &generic_types() const
+  {
+    return (const generic_typest &)(find(ID_generic_types).get_sub());
```
can you use `static_cast` instead?

thomasspriggs replied:
No, putting a static cast here causes a compile error.

romainbrenguier commented on src/java_bytecode/java_types.h:
```diff
@@ -518,6 +518,82 @@ to_java_specialized_generic_class_type(typet &type)
+
+  generic_typest &generic_types()
+  {
+    return (generic_typest &)(add(ID_generic_types).get_sub());
```
I was confused by the name of the function `add`, but yes that this is correct...

Commit:
Correcting generic parameters in bases for implicitly generic classes

If the superclass or interface uses a generic parameter inherited from
an outer class, make sure that the parameter for the base points to the
parameter of the outer class.

thk123 commented:
CI failure looks like it might be nonsense so restarted it, is this ready for a re-review @jeannielynnmoulton ?

majakusber commented:
@here This is now ready for re-review. 

In order to fix failing CI, we went back to the original design for the new type for a generic base which is `java_generic_symbol_typet` that extends `symbol_typet` (used for non-generic bases). The advantage is that if the fact that the base is generic is for whatever reason ignored at any point, nothing will break (this works generally for generics in our code as well as in Java itself).

thk123 commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -85,6 +85,58 @@ class java_bytecode_convert_classt:public messaget
+/// \return Reference of the generic superclass, or empty if the superclass
+/// is not generic
+static std::string
+extract_generic_superclass_reference(const std::string &signature)
```
Consider using `optionalt<std::string>` rather than the empty string to denote the non-generic case.

chrisr-diffblue replied:
I'm a bit out of the loop on the generics stuff now :-) so for people like me, it might also be nice if the Doxygen comments for these functions included an example of an input signature string, and an example of what the returned string looks like.

thk123 commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -85,6 +85,58 @@ class java_bytecode_convert_classt:public messaget
+/// \param interface_name The interface name
+/// \return Reference of the generic interface, or empty if the interface
+/// is not generic
+static std::string extract_generic_interface_reference(
```
Again - I think `optional` would be clearer here.

thk123 commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -145,10 +197,28 @@ void java_bytecode_convert_classt::convert(const classt &c)
+    }
     class_typet::componentt base_class_field;
-    base_class_field.type()=base;
+    base_class_field.type() = class_type.bases().at(0).type();
```
How did this work if `base` was an `exprt` and `type()` was a `typet`?

majakusber replied:
`base` was a `symbol_typet`, only once called `.add_base(base)` it was wrapped into an `exprt`.

thk123 commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -158,8 +228,30 @@ void java_bytecode_convert_classt::convert(const classt &c)
+    // LGenericInterface<LInteger;>;'
+    if(
+      c.signature.has_value() &&
+      !extract_generic_interface_reference(
```
Rather than calling the function twice it is probably clearer to store the result. It might also simplify things if you shift the check for `c.signature.has_value()`  inside the extraction method

thk123 commented on unit/java_bytecode/java_bytecode_parse_generics/InterfaceGeneric.java:
```diff
@@ -0,0 +1,4 @@
+interface InterfaceGeneric<E>
+{
+    E someMethod();
+}
```
Missing new line

thk123 reviewed (commented):
Not yet reviewed the unit tests but suggestions on code, though nothing blocking.

thk123 commented on unit/java_bytecode/java_bytecode_parse_generics/DerivedGenerics.java:
```diff
@@ -0,0 +1,158 @@
+    int field;
+}
+
+class ExtendsNotGeneric extends NotGeneric {
```
This doesn't seem to be involved?

thk123 commented on unit/java_bytecode/java_bytecode_parse_generics/DerivedGenerics.java:
```diff
@@ -0,0 +1,158 @@
+
+}
+
+class ImplementsInterface implements Interface {
```
This doesn't appear to be used anywhere

thk123 reviewed (approved):
These unit tests are really good :+1:

chrisr-diffblue commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -85,6 +85,58 @@ class java_bytecode_convert_classt:public messaget
+{
+  if(signature.find("L" + interface_name + "<") != std::string::npos)
+  {
+    const size_t start = signature.find("L" + interface_name + "<");
```
Would it be better to move this `start` variable and its expression outside the `if()` statement and only do the `find` once?

majakusber commented:
@thk123 @chrisr-diffblue Thanks for reviews, I updated the code correspondingly. Will update the test-gen pointer bump shortly https://github.com/diffblue/test-gen/pull/1401.

thk123 commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -86,55 +86,67 @@ class java_bytecode_convert_classt:public messaget
+    // extract the superclass reference
+    const std::string superclass_ref = signature.value().substr(
+      start,
+      find_closing_semi_colon_for_reference_type(signature.value(), start) -
```
Nit: the minus trailing at the end is a little hard to read, I'd prefer pulling `find_closing_semi_colon_for_reference_type(signature.value(), start) ` into a variable if needed to get param on one line

chrisr-diffblue reviewed (approved):
Looks good to me too. It would probably be nicer if you can squash the last four commits into their appropriate base commits before merging, just to keep the commit history cleaner, but other than that it seems good to go :-)

smowton commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -85,6 +85,70 @@ class java_bytecode_convert_classt:public messaget
+/// Example:
+/// - class: A<T> extends B<T, Integer> implements C, D<T>
+/// - signature: <T:Ljava/lang/Object;>B<TT;Ljava/lang/Integer;>;LC;LD<TT;>;
+/// - superclass reference: B<TT;Ljava/lang/Integer;>;
```
"returned superclass reference"?

smowton commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -85,6 +85,70 @@ class java_bytecode_convert_classt:public messaget
+
+    // if the superclass is generic then the reference is of form
+    // Lsuperclass-name<generic-types;>;
+    if(superclass_ref.substr(superclass_ref.length() - 2) == ">;")
```
Use `has_suffix`

smowton commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -85,6 +85,70 @@ class java_bytecode_convert_classt:public messaget
+{
+  if(signature.has_value())
+  {
+    const size_t start = signature.value().find("L" + interface_name + "<");
```
Could the interface also occur earlier? For example, for `class A extends ArrayList<Closable> implements Closable`, would this find the reference to Closable as an argument to ArrayList rather than the interface as desired?

majakusber replied:
Thank you for pointing this out! Yes, the interface can occur earlier as in the example you gave. However, we are only interested in the reference to the interface if it is generic which is ensured by looking for substring with `<` after the name of the interface. In your example there would be no problem. The problem arises in case like this `class A extends B<Interface<String>> implements Interface<Integer>` where the signature is:
`LB<LInterface<Ljava/lang/String;>;>;LInterface<Ljava/lang/Integer;>;`
The function should return reference `LInterface<Ljava/lang/Integer;>;` but this implementation returns `LInterface<Ljava/lang/String;>;`. I've corrected the code and added unit tests for this case.

majakusber replied:
And this also added unit tests for nested generics in bases which we didn't have before so double thanks :)

smowton commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -85,6 +85,70 @@ class java_bytecode_convert_classt:public messaget
+    if(start != std::string::npos)
+      return signature.value().substr(
+        start,
+        find_closing_semi_colon_for_reference_type(signature.value(), start) -
```
Suggest parentheses to clarify whether you intend 1 more or 1 less (i.e. a - (b + 1) vs. (a - b) + 1)

smowton commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -145,10 +209,26 @@ void java_bytecode_convert_classt::convert(const classt &c)
+      extract_generic_superclass_reference(c.signature);
+    if(superclass_ref.has_value())
+    {
+      const java_generic_symbol_typet gen_base(
```
generic_base? `gen_base` suggests "generated base" to me

smowton commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -158,8 +238,24 @@ void java_bytecode_convert_classt::convert(const classt &c)
+      extract_generic_interface_reference(c.signature, id2string(interface));
+    if(interface_ref.has_value())
+    {
+      const java_generic_symbol_typet gen_base(
```
Ditto

smowton reviewed (commented):
Mostly looks fine, but please have a look at comment "Could the interface also occur earlier..." before I approve

thk123 commented:
This requires approval from @smowton due to modification of `irep_ids.def` - I've created a PR to exclude this file from the high risk section of the code owners - #1791
